### PR TITLE
DTrace updates for showing Upstairs state

### PR DESCRIPTION
**Pull request by leftwo**
First pass at using DTrace to display some internal Upstairs state, showing
the internal state of the three downstairs as well as a count for the
active jobs of both the guest side and the downstairs side.

Updated comments in the DTrace scripts.

Moved the older DTrace scripts into the newer attribute macro instead
of the build.rs script.  Removed the old build.rs script and d file.

### CODE REVIEW

File: upstairs/src/lib.rs [hunk 1]
```diff
@@ -3706,6 +3746,27 @@ async fn process_new_io(
...
+ * Update the counters used by dtrace probes.
+ * This one method will update the fields of the
+ * up_status counter.
+ */
+fn stat_update(up: &Arc<Upstairs>, msg: &str) {
+    let arg = Arg {
+        up_count: up.up_work_active(),
+        ds_count: up.ds_work_active(),
+        buffer: up.ds_state_copy(),
+    };
```
[the remark this answers is not in the document]
Comment (leftwo) [reply upstairs/src/lib.rs]: Your suggestion to use Bens suggestion has worked here, changed in 51af15d.

File: tools/README.md [hunk 2]
```diff
@@ -16,20 +16,42 @@ If a downstairs dies for any other reason then being killed with the
...
 
-You start crucible, then run the above script.  When you are ready to see
-results, hit Control-C and the final counts will be printed.
+When you are ready to see results, hit Control-C and the final counts will
+be printed.
 
 Here is an example of how it might look:
 ```
-final:crucible alan$ sudo dtrace -s tools/tracegw.d
+final:crucible alan$ sudo dtrace Z -s tools/tracegw.d
```
[the remark this answers is not in the document]
Comment (leftwo) [reply tools/README.md]: Fixed in 51af15d

File: tools/perfgw.d [hunk 3]
```diff
@@ -1,23 +1,23 @@
...
+ * Trace all the guest submitted and completed IOs.
+ */
+cdt*:::gw_read_start,
+cdt*:::gw_write_start,
+cdt*:::gw_flush_start
 {
     start[arg0] = timestamp;
 }
-crutrace*:::gw_read_end
+cdt*:::gw_read_end
```
[the remark this answers is not in the document]
Comment (leftwo) [reply tools/perfgw.d:10]: Good find, fixed in 51af15d

File: upstairs/src/lib.rs [hunk 4]
```diff
@@ -2237,6 +2247,35 @@ impl Upstairs {
...
+     * Return a copy of the DsState vec.
+     * DTraces uses this.
+     */
+    fn ds_state_copy(&self) -> Vec<DsState> {
+        let mut buffer: Vec<DsState> = Vec::new();
+        let ds = self.downstairs.lock().unwrap();
+        for (index, dst) in ds.ds_state.iter().clone().enumerate() {
+            buffer.insert(index, *dst);
+        }
+        buffer
```
[the remark this answers is not in the document]
Comment (leftwo) [reply upstairs/src/lib.rs]: You are correct.  I tried a few other ways that did not work, but this one did.
Fixed in 51af15d

File: upstairs/src/lib.rs [hunk 5]
```diff
@@ -38,8 +39,17 @@ mod test;
 
 pub use pseudo_file::CruciblePseudoFile;
 
-// Include the Rust implementation generated by the build script.
-include!(concat!(env!("OUT_DIR"), "/crutrace.rs"));
+#[usdt::provider]
+mod cdt {
+    use crate::Arg;
+    fn up_status(_: String, arg: &Arg) {}
```
[the remark this answers is not in the document]
Comment (leftwo) [reply upstairs/src/lib.rs]: Yep, that worked like a charm.
Fixed in 51af15d

**Comment by leftwo** [comment]
I should have all comments addressed now with 51af15d.
If there is anything else, let me know.

I have a bunch more ideas on how to build on this, but I'll save them for the next
PR and build on this base.

File: tools/perfgw.d [hunk 6]
```diff
@@ -1,23 +1,23 @@
...
 {
     @time["read"] = quantize(timestamp - start[arg0]);
     start[arg0] = 0;
 }
-crutrace*:::gw_write_end
+cdt*:::gw_write_end
```
[the remark this answers is not in the document]
Comment (leftwo) [reply tools/perfgw.d:16]: Fixed in 51af15d

File: upstairs/src/lib.rs [hunk 7]
```diff
@@ -3706,6 +3741,29 @@ async fn process_new_io(
...
+    ds_count: u32,
+    ds_state: Vec<DsState>,
+}
+
+/**
+ * Update the counters used by dtrace probes.
+ * This one method will update the fields of the
+ * up_status counter.
+ */
+fn stat_update(up: &Arc<Upstairs>, msg: &str) {
```
Comment (bnaecker) [inline upstairs/src/lib.rs:3756]: Not sure if it matters at this point, but you _could_ further reduce the likelihood of a disabled-probe effect by either inlining this function, or calling the macro directly in place of calls to this function. Totally up to you, just a thought.

**Review by bnaecker — APPROVED** [review]

File: upstairs/src/lib.rs (hunk 7 above)
Comment (leftwo) [reply upstairs/src/lib.rs:3757]: That is a good idea.  I'm going to be adding a bunch more stats, and I think I'll be
throwing this function out and sprinkling things elsewhere.  I wanted to make a first
pass at this to make sure I had the right idea how to do it.

**Pull request merged**